e2e/tests/update_params_test.go

```diff
@@ -0,0 +1,583 @@
+		paramsMap[paramName] = paramValueType
+	}
+
+	return paramsMap
+}
+
+func (s *suite) parseParam(table gocuke.DataTable, rowIdx int) (paramName string, paramValueType anyMap) {
+	var paramValue any
+	paramName = table.Cell(rowIdx, 0).String()
+	paramType := table.Cell(rowIdx, 2).String()
+
+	switch paramType {
+	case "string":
+		paramValue = table.Cell(rowIdx, 1).String()
+	case "int64":
+		paramValue = table.Cell(rowIdx, 1).Int64()
+	case "bytes":
+		paramValue = []byte(table.Cell(rowIdx, 1).String())
+	default:
+		s.Fatalf("unexpected param type %q", paramType)
+	}
+
+	return paramName, anyMap{
+		paramType:  paramType,
+		paramValue: paramValue,
+	}
+}
+
+func (s *suite) paramsMapToMsgUpdateParams(moduleName string, paramsMap map[string]anyMap) (msg cosmostypes.Msg) {
+	authority := authtypes.NewModuleAddress(s.granterName).String()
+
+	switch moduleName {
+	case tokenomicstypes.ModuleName:
+		msgUpdateParams := &tokenomicstypes.MsgUpdateParams{
+			Authority: authority,
+			Params:    tokenomicstypes.Params{},
+		}
+
+		for paramName, paramValue := range paramsMap {
+			//s.Logf("paramName: %s, paramValue: %v", paramName, paramValue.paramValue)
+			switch paramName {
+			case "compute_units_to_tokens_multiplier":
+				msgUpdateParams.Params.ComputeUnitsToTokensMultiplier = uint64(paramValue.paramValue.(int64))
+			default:
+				s.Fatalf("unexpected %q type param name %q", paramValue.paramType, paramName)
+			}
+		}
+		msg = proto.Message(msgUpdateParams)
+
+	case prooftypes.ModuleName:
+		msgUpdateParams := &prooftypes.MsgUpdateParams{
+			Authority: authority,
+			Params:    prooftypes.Params{},
+		}
+
+		for paramName, paramValue := range paramsMap {
+			s.Logf("paramName: %s, paramValue: %v", paramName, paramValue.paramValue)
+			switch paramName {
+			case "min_relay_difficulty_bits":
+				msgUpdateParams.Params.MinRelayDifficultyBits = uint64(paramValue.paramValue.(int64))
+			default:
+				s.Fatalf("unexpected %q type param name %q", paramValue.paramType, paramName)
+			}
+		}
+		msg = proto.Message(msgUpdateParams)
+
+	default:
+		err := fmt.Errorf("unexpected module name %q", moduleName)
+		s.Fatal(err)
+		panic(err)
+	}
+
+	return msg
+}
+
+// TODO_IN_THIS_COMMIT: refactor with `#paramsMapToMsgUpdateParams`.
+func (s *suite) paramsMapToMsgUpdateParam(moduleName, paramName string, paramTypeValue anyMap) (msg cosmostypes.Msg) {
+	authority := authtypes.NewModuleAddress(s.granterName).String()
+
+	//switch paramTypeValue.paramType {
+	//case "string":
+	//	msg = paramMsgUpdateFromParam[tokenomicstypes.MsgUpdateParam_AsString](
+	//		moduleName,
+	//		authority,
+	//		paramName,
+	//		paramTypeValue,
+	//	)
+	//default:
+	//	s.Fatalf("unexpected param type %q", paramTypeValue.paramType)
+	//}
+
+	// TODO_IMPROVE: refactor/simplify?
+	switch moduleName {
+	case tokenomicstypes.ModuleName:
+		//msg := &tokenomicstypes.MsgUpdateParam{
+		//	Authority: authority,
+		//	Name:      paramName,
+		//	AsType: paramAsType(paramTypeValue),
```

Comment:
Do you intend to make `paramAsType` work? What's blocking from having it?

Reply:
I've tried a few different approaches to work with the `AsType` value as an interface but go just doesn't want to do it. It doesn't like that the interface referenced within the .pb.go files are not exported (`isMsgUpdateParam_AsType`).